### ISSUE TITLE
Add GPGPU

### ISSUE THREAD

**Issue by querielo** [playcanvas/engine#4769]
Problem: I experimented [with water surface simulation](https://playcanv.as/b/1a320545), and found that there is no library for GPGPU in Playcanvas. So, a user has to create an environment for that themselves.

**Suggestion**: add GPUCompute renderer that can be used as GPGPU.
**Primitives**: 
1) variables are textures
2) the computation is based on a fragment shader 
  a) it uses textures as variables
  b) it happens before frame rendering
  c) a result of the computation is texture (another variable)
  d) a variable can have itself as a dependency (!)

**Comment by kungfooman** [comment]
So you want PlayCanvas to implement basically something like https://github.com/gpujs/gpu.js?

Is your demo open source, so we can see the code of what you are currently doing to achieve the water splash effect? I only see the link to your running example.

Since GPU's become more and more powerful, I think this opens all kinds of doors, but I'm still confused what changes would be required. Right now gpu.js is a 2nd class citizen and you ask it to be 1st class in PlayCanvas?

**Comment by querielo** [comment]
> So you want PlayCanvas to implement basically something like https://github.com/gpujs/gpu.js?

Yes, or maybe something simplier. At least I don't understand how to pass result of computation from gpu.js to Playcanvas Texture without CPU<->GPU round trip.


> Is your demo open source, so we can see the code of what you are currently doing to achieve the water splash effect? I only see the link to your running example.

I need to polish the code and improve naming. Maybe I do it later. But you can find idea here:
<img width="734" alt="Screen Shot 2022-10-31 at 22 28 45" src="https://user-images.githubusercontent.com/104348270/199082999-986368ff-067f-4999-84dc-98e53f838734.png">

**Comment by LeXXik** [comment]
@querielo I've made one exactly like you want like a year ago or so. Each variable is a texture, and supports itself as a dependency. The only reason I haven't published it yet, is I wasn't able to get my hands on some nice example that would make use if it. I was thinking to use boids, and move a swarm around. Made a simple butterfly animated in vertex shader for it. Wanted to make a bunch of them fly around a flower or something. Still in my todo list since then :D If you have a good and simple shader example, we could make a public project together.

**Comment by querielo** [comment]
@LeXXik  I thought about particles that visualize disintegrated objects that has to be hidden (like dead players or inactive game objects). So instead of hiding objects in one frame the objects are dissolved with random walking particles.
Or we could do magic stuff like fireballs, lightning, firework :-)

It can help to develop a suitable GPGPU subsystem. I open for everything.

**Comment by LeXXik** [comment]
I have a project here with the gpgpu, but it needs a better example. At the moment, there is just a lone butterfly flying away :) But it shows how to register variables and add self-dependency. 

https://playcanvas.com/project/805322/overview/gpgpu

Drop me a DM on the forums with your username, and I will add you to the project:
https://forum.playcanvas.com/u/LeXXik

**Comment by querielo** [comment]
@LeXXik I am a newbie on the forum. So, I don't know how to do it :-)

> You don’t have any messages
Need to have a direct personal conversation with someone, outside the normal conversational flow? Message them by selecting their avatar and using the  message button.
>
> If you need help, you can [message a staff member](https://forum.playcanvas.com/about).

But I can't find the message button.

<img width="637" alt="Screen Shot 2022-11-11 at 10 52 46" src="https://user-images.githubusercontent.com/104348270/201280953-f131d1fb-2abc-4a54-a3d5-e56ac525d57c.png">
<img width="362" alt="Screen Shot 2022-11-11 at 10 53 21" src="https://user-images.githubusercontent.com/104348270/201281042-acd55568-139e-4acb-8f09-788e80828f3f.png">

Could you write to my email? (see my GitHub account)

**Comment by mvaligursky** [comment]
Perhaps this could be used to deliver GPGPU https://github.com/amandaghassaei/gpu-io

**Comment by mvaligursky** [comment]
I'm making a call that this can easily be implemented on top of the engine, and there is no need for this to be part of the code engine. Closing for now.
